### PR TITLE
Update CredentialsCache to handle OnBehalfof credentials

### PR DESCRIPTION
**Pull request by richma-ms**
This change updates the CredentialsCache and PersistedCredentialsCache so that we maintain a correct mapping of OnBehalfOf token credentials.

For OnBehalfOf connections, everything we receive on the CbsNode points to the acting child EdgeHub as the owner of the credentials.  But adding just this to the cache would be incorrect, as the individual links are opened OnBehalfOf a descendent.  The auth-chain that we receive when these links open contains the actual auth target, so the cache needs remember map the auth-target to the token that we received earlier for the actor EdgeHub.  This way when the reauthenticator iterates over the existing OnBehalfOf connections, the credentials cache can surface up the actor's token for validation.

Additionally, when the actor EdgeHub token is updated, the credentials cache now needs to go update all the corresponding entries that the EdgeHub is acting OnBehalfOf.

### CODE REVIEW

File: edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.Core/NestedCredentialCache.cs [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+
+                return Option.Some(clientCredentials);
+            }
+        }
+
+        void UpdateCachedCredentials(IClientCredentials clientCredentials)
+        {
+            string targetId = AuthChainHelpers.GetAuthTarget(clientCredentials.AuthChain).GetOrElse(clientCredentials.Identity.Id);
+
+            if (clientCredentials.AuthChain.HasValue)
```
Comment (varunpuranik) [inline edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.Core/NestedCredentialCache.cs:59]: better way to do this is clientCredentials.AuthChain.Map().OrElse(), or clientCredentials.AuthChain.Match()

File: edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.Core/NestedCredentialCache.cs [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+                        this.credentials[client] = updatedCredentials;
+                    }
+                }
+            }
+
+            // Update the client credentials
+            this.credentials[targetId] = clientCredentials;
+        }
+
+        public static IClientCredentials GetCredentialsWithOnBehalfOfUpdates(IClientCredentials originalClientCreds, IClientCredentials updatedActorCreds)
```
Comment (varunpuranik) [inline edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.Core/NestedCredentialCache.cs:96]: Make it private (or internal if used in tests)

File: edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.Core/PersistedTokenCredentialsCache.cs [hunk 3]
```diff
@@ -22,10 +23,13 @@ public async Task Add(IClientCredentials clientCredentials)
...
             if (clientCredentials is ITokenCredentials tokenCredentials)
             {
+                string targetId = AuthChainHelpers.GetAuthTarget(tokenCredentials.AuthChain).GetOrElse(tokenCredentials.Identity.Id);
+
                 try
                 {
-                    var tokenCredentialsData = new TokenCredentialsData(tokenCredentials.Token, tokenCredentials.IsUpdatable);
-                    await this.encryptedStore.Put(tokenCredentials.Identity.Id, tokenCredentialsData.ToJson());
+                    var tokenCredentialsData = new TokenCredentialsData(tokenCredentials.Token, tokenCredentials.IsUpdatable, tokenCredentials.AuthChain);
+                    string tokenCredentialsString = JsonConvert.SerializeObject(tokenCredentialsData);
```
Comment (varunpuranik) [inline edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.Core/PersistedTokenCredentialsCache.cs:31]: Why does toJson not work here?

**Review by varunpuranik — APPROVED** [review]

File: edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.Core/PersistedTokenCredentialsCache.cs (hunk 3 above)
Comment (richma-ms) [reply edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.Core/PersistedTokenCredentialsCache.cs:31]: there's a definition ambiguity where ToJson() is declare in Newtonsoft.Json and Edge.Util.Json, this is just my work around for it.

File: edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.Core/NestedCredentialCache.cs (hunk 1 above)
Comment (richma-ms) [reply edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.Core/NestedCredentialCache.cs:59]: both approach requires a return value in the lambdas, which doesn't really apply to either if/else block.  If I make them return some arbitrary type then it prob looks more awkward =(

**Pull request merged**